libcontainer/state_linux.go

```diff
@@ -48,6 +49,14 @@ func destroy(c *Container) error {
 		// Likely to fail when c.config.RootlessCgroups is true
 		_ = signalAllProcesses(c.cgroupManager, unix.SIGKILL)
 	}
+
+	// Likely to fail if the init process no longer exist.
```

Comment:
I only found this place to hook for restoring these devices, but since the start process may not exist , when the network namespace is not provided it is not able to find the network namespace ... please advice

Reply:
any point where we can get the network namespace and the initial process has not started yet

Reply:
What do you refer to with "the initial process" ? The container process? Or the runc init processes?

All runc code runs before the container process, once we exec the container runc is out of the picture.

Assuming you mean the container process, the namespaces are created in nsexec.c, there are a lot of things done after that in go too. Like the mountToRootfs(), etc.

I have not look in detail to suggest where to do it, but take a look and just add it somewhere. People will review and it can be moved in the worst case :)

Reply:
@aojea ^

Reply:
> I don't know when CNI is being called by upper container runtimes (like containerd or crio), but I guess we should move the interface at that point (whatever it is, create or start)

in containerd the network namespace is created and the CNI is called to provide IPs and the network interface, once that is set the containers starts to be created

Reply:
the setup seems to be in the correct place now, before the Prestart hook

Reply:
per latest conversation, we delegate the lifecycle of the interface to the network namespace owner, and if nothing happens the kernel will handle the cleanup ... this is less confusing than having different behaviors on the cases of process crash and virtual and physical interfaces cleanup